Support Python SDK 3.0.0

antonpirker opened the pull request:
Update the codebase so it works with `Sentry Python SDK 3.0.0 alpha`.

Sentry SDK 3.0.0 has some backwards incompatibilities and uses Opentelementry under the hood to create spans. 

### What this PR does:
- replace `sentry_sdk.start_transaction()` with `sentry_sdk.start_span()`
- replace `span.set_data()` with `span.set_attribute()`
- replace `scope.transaction` with `scope.root_span`
- replace `Span.containing_transaction` with `Span.root_span`
- replace `custom_sampling_context` with `attributes` parameter
- replace `sentry_sdk.scope.get_*_scope()` with `sentry_sdk.get_*_scope()`
- replace one `scope.add_attachment()` with `sentry_sdk.add_attachment()`
- change one `continue_trace()` to the new context manager.
- update `traces_sampler` to use new `sampling_context` format
- added a few tests

See also the migration guide:
https://docs.sentry.io/platforms/python/migration/2.x-to-3.x

---


### Related, but split out into separate PRs:
- make sure `traces_sampler` works the same as before (check for `sample_rate`): https://github.com/getsentry/sentry/pull/93011
- make **all** `get_isolation_scope` calls use the top-level API. Currently, some calls where we `from sentry_sdk import Scope` then call `Scope.get_isolation_scope` still need to be converted to the top-level API: https://github.com/getsentry/sentry/pull/93307
- check `event["measurements"]` places if those can be removed (because SDK does not support measurements anymore): https://github.com/getsentry/sentry/pull/92718
- change set_span_data() to set_span_attribute() but check for value NOT be a dict: #92946
- replace `with sentry_sdk.init(dsn=...)` with `with sentry_sdk.new_scope() as scope: scope.set_client(sentry_sdk.Client(dsn=...))` in `src/sentry/runner/main.py`: #92944
- remove `propagate_hub=True` from `ThreadingIntegration`: #93016
- remove minimetrics from the sentry codebase, because the metrics product was never released: #93595
- https://github.com/getsentry/getsentry/pull/17553
- https://github.com/getsentry/devenv/pull/194
- https://github.com/getsentry/devservices/pull/277
- https://github.com/getsentry/pytest-sentry/pull/39
- https://github.com/getsentry/sentry-docs/pull/13939

antonpirker committed:
Update requirements

antonpirker committed:
Remove propagate_hub because it is not needed because its the default behavior now

antonpirker committed:
Removed metrics_noop because there are no metrics anymore and thus not needed

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Make tests run locally until pytest-sentry is 3.0 compatible

antonpirker committed:
Replace .set_data with .set_attribute

antonpirker committed:
changed add_attachment

antonpirker committed:
Changed start_transaction to start_span and custom_sampling_context to attributes

antonpirker committed:
Changed continue_trace to use context manager

antonpirker committed:
Removed set_measurements helper function because it is not used in codebase and also gone in SDK 3.0

antonpirker committed:
oops

antonpirker committed:
marked places where dict-attributes are

antonpirker committed:
Replaced  with

antonpirker committed:
one more comment

antonpirker committed:
more comments

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
bump pytest-sentry

getsantry[bot] committed:
:snowflake: re-freeze requirements

antonpirker committed:
trigger ci

antonpirker committed:
update pytest-sentry

antonpirker committed:
Replaced .transaction with .root_span

antonpirker committed:
Always use top level api to get scope

antonpirker committed:
use sample_rate in attributes instead of custom_sampling_context

antonpirker committed:
trigger ci

antonpirker committed:
Replaced .containing_transaction with .root_span

antonpirker committed:
Removed unused option

antonpirker committed:
set num_of_spanes in event.data not event.measurement

antonpirker committed:
fixed linting

antonpirker committed:
Ignore measurements mypy check (because measuerements have already been removed from sentry sdk)

antonpirker committed:
replaced  with

antonpirker committed:
formatting

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
does this work?

antonpirker committed:
restart ci

antonpirker committed:
Updated assert

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
trigger ci

antonpirker committed:
Renamed set_span_data to set_span_attribute to reflect new api

antonpirker committed:
Reverted attribute name change. this is safer

antonpirker committed:
testing something

antonpirker committed:
testing something

antonpirker committed:
test

antonpirker committed:
trigger ci

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
cleanup

asottile-sentry commented:
would it be possible to split some of this out to separate PRs so it's more easily reviewable?

szokeasaurusrex committed:
test: Remove redundant assertions

These assertions are redundant. We already assert that `response.data` has the correct contents in all three tests. If we want to check `error.to_public_dict()`'s value, we should do so in a separate unit test.

Hopefully unblocks #92011

szokeasaurusrex committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

szokeasaurusrex committed:
Merge branch 'szokeasaurusrex/remove-redundant-assertion' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Those set_attribute calls are fine. See https://github.com/getsentry/sentry/pull/92720

szokeasaurusrex committed:
ref: Use `set_client` to change DSN

`with sentry_sdk.init` is deprecated and will be removed in Sentry SDK 3.0.

Split off from #92011

szokeasaurusrex committed:
ref: Rename `set_span_data` to `set_span_attribute`

`sentry_sdk.Span.set_data` will be replaced with `sentry_sdk.Span.set_attribute` in Sentry SDK 3.0.

This change renames the `set_span_data` method to `set_span_attribute` to match the new method name in the SDK.

Split off from #92011

szokeasaurusrex committed:
Merge branch 'szokeasaurusrex/set_span_attribute' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'antonpirker/code-changes-for-sdk-3.0.0' of https://github.com/getsentry/sentry into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

szokeasaurusrex committed:
ref: Remove `propagate_hub` to `ThreadingIntegration`

This parameter is deprecated as of Sentry SDK 2.0.0; the default value is now `True`, so passing this parameter is no longer needed.

szokeasaurusrex committed:
ref(sdk): Use top-level `get_current_scope`

Using `sentry_sdk.Scope.get_current_scope` has been deprecated in favor of `sentry_sdk.get_current_scope`.

Going forward, we generally would like SDK users to stick to the top-level APIs.

Split off from #92011.

szokeasaurusrex committed:
ref(sdk): Migrate to `get_isolation_scope`

Migrate calls of `sentry_sdk.Scope.get_isolation_scope` to `sentry_sdk.get_isolation_scope`. Calls where `Scope` is first imported and then where `Scope.get_isolation_scope` is called are left as-is; we will address those separately.

Split off from #92011.

szokeasaurusrex committed:
ref(sdk): Migrate to `sentry_sdk.get_global_scope`

Migrate calls of `sentry_sdk.Scope.get_global_scope` to `sentry_sdk.get_global_scope`.

Split off from #92011.

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

szokeasaurusrex committed:
Merge branch 'szokeasaurusrex/get_global_scope' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'antonpirker/code-changes-for-sdk-3.0.0' of https://github.com/getsentry/sentry into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Removed debug code

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

szokeasaurusrex committed:
test: Add test for custom sample rates (#93011)

<!-- Describe your PR here. -->

<!--

  Sentry employees and contractors can delete or ignore the following.

-->

### Legal Boilerplate

Look, I get it. The entity doing business as "Sentry" was incorporated
in the State of Delaware in 2015 as Functional Software, Inc. and is
gonna need some rights from me in order to utilize my contributions in
this here PR. So here's the deal: I retain all rights, title and
interest in and to my contributions, and by keeping this boilerplate
intact I confirm that Sentry can use, modify, copy, and redistribute my
contributions, under Sentry's choice of terms.

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
set_data -> set_attribute

antonpirker committed:
trigger ci

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
new pytest-sentry

antonpirker committed:
update

antonpirker committed:
client is not an argument of Scope constructor anymore

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

untitaker commented:
> there seems to be a lot of unrelated changes in this PR?


are there? some new sdk-related tests, other than that it's a mass-replace

sl0thentr0py reviewed (changes requested):
@antonpirker 
this is not ready to merge, we have to actually update the `traces_sampler` since we no longer have the same stuff in `sampling_context`

https://github.com/getsentry/sentry/blob/f5fe6b08fe5f654fc0f94c2708485d748c61be96/src/sentry/utils/sdk.py#L180-L201

see here for the migration
https://docs.sentry.io/platforms/python/migration/2.x-to-3.x#sampling

sentrivana committed:
Migrate sampling_context to use SDK 3.0 attributes (#94022)

See the Celery and WSGI sections [in the migration
guide](https://docs.sentry.io/platforms/python/migration/2.x-to-3.x#sampling)

Closes https://github.com/getsentry/sentry-python/issues/4505

sentrivana commented:
For posterity, the issue from [Neel's comment above](https://github.com/getsentry/sentry/pull/92011#pullrequestreview-2945894526) was fixed [here](https://github.com/getsentry/sentry/pull/92011/commits/25712df985d31a083cebc617b6064adcbf8b5730)

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

sl0thentr0py reviewed (approved):
stamping

antonpirker committed:
naming

antonpirker committed:
test parametrization

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

asottile-sentry commented on requirements-dev-frozen.txt:
```diff
@@ -82,6 +83,7 @@ httpx==0.25.2
 hyperframe==6.1.0
 identify==2.6.1
 idna==3.7
+importlib-metadata==8.6.1
```
not to mention the package has been removed from a bunch of operating system vendors so it would prevent us from being packaged there

sentrivana replied:
Agreed that it'd be great to get rid of it but not likely to happen, see https://github.com/open-telemetry/opentelemetry-python/issues/3234#issuecomment-2425870153 & https://github.com/open-telemetry/opentelemetry-python/pull/4177

sl0thentr0py replied:
@sentrivana since you have more context, can you summarize **why** they are stuck with `importlib-metadata`? I see it has to do with the entrypoints logic but I don't understand why that should belong to the core `opentelemetry-api` package and not the `-contrib` stuff that actually instruments.

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

antonpirker committed:
Removed new tests to test something

antonpirker committed:
small fix

antonpirker committed:
Revert "Removed new tests to test something"

This reverts commit 3674af78b24ca9a9577d55ad5d2b165658ce31d5.

antonpirker committed:
better checks

antonpirker committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

untitaker commented:
> this means some background thread started a database connection in some test (it's pretty hard to actually narrow down). it's likely caused by this PR unfortunately :(

I think this flake also exists on main. @joshuarli tagged me in a flaky test ticket that had the same issue.

sentrivana committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

asottile-sentry commented:
> > this means some background thread started a database connection in some test (it's pretty hard to actually narrow down). it's likely caused by this PR unfortunately :(
> 
> I think this flake also exists on main. @joshuarli tagged me in a flaky test ticket that had the same issue.

it's a bit flaky on mainline yeah -- but not to the level it is failing here

sl0thentr0py committed:
Fix scope level calls

sl0thentr0py committed:
Merge remote-tracking branch 'origin/master' into antonpirker/code-changes-for-sdk-3.0.0

sl0thentr0py committed:
Bump pytest-sentry to 0.5.1

sl0thentr0py commented:
only one test is failing now after the `pytest-sentry` update, I _think_ that substantially reduced the reports here
https://sentry.sentry.io/issues/?project=2423079&query=is%3Aunresolved%20pytest_environ.GITHUB_REF%3Arefs%2Fpull%2F92011%2Fmerge&referrer=issue-list&statsPeriod=1h
but there's still a few of them that I think are valid? please correct me if I'm wrong!

sentrivana committed:
Merge branch 'master' into antonpirker/code-changes-for-sdk-3.0.0

sentrivana commented:
Reverted because a change in issue grouping introduced by this release backlogged canary. We'll fix this and dogfood again next week.
